utils/zerovec/derive/src/ule.rs

```diff
@@ -77,6 +81,248 @@ pub fn derive_impl(input: &DeriveInput) -> TokenStream2 {
     }
 }
 
+pub fn make_ule_impl(attr: AttributeArgs, input: DeriveInput) -> TokenStream2 {
+    if input.generics.type_params().next().is_some()
+        || input.generics.lifetimes().next().is_some()
+        || input.generics.const_params().next().is_some()
+    {
+        return Error::new(
+            input.generics.span(),
+            "derive(ULE) must be applied to a struct without any generics",
+        )
+        .to_compile_error();
+    }
+
+    if !attr.is_empty() {
+        return Error::new(
+            input.span(),
+            "#[make_ule] does not currently support any arguments",
+        )
+        .to_compile_error();
+    }
+    let name = &input.ident;
+    let ule_name = Ident::new(&format!("{}ULE", name), Span::call_site());
+
+    let ule_stuff = match input.data {
+        Data::Struct(ref s) => make_ule_struct_impl(name, &ule_name, &input, s),
+        Data::Enum(ref e) => make_ule_enum_impl(name, &ule_name, &input, e),
+        _ => {
+            return Error::new(input.span(), "#[make_ule] must be applied to a struct")
+                .to_compile_error();
+        }
+    };
+
+    // Todo: opt-out for ZMKV impl
+
+    quote!(
+        #input
+
+        #ule_stuff
+
+        impl<'a> zerovec::map::ZeroMapKV<'a> for #name {
+            type Container = zerovec::ZeroVec<'a, #name>;
+            type GetType = #ule_name;
+            type OwnedType = #name;
+        }
+    )
+}
+
+fn make_ule_enum_impl(
+    name: &Ident,
+    ule_name: &Ident,
+    input: &DeriveInput,
+    enu: &DataEnum,
+) -> TokenStream2 {
+    // We could support more int reprs in the future if needed
+    if !utils::has_valid_repr(&input.attrs, |r| r == "u8") {
+        return Error::new(
+            input.span(),
+            "#[make_ule] can only be applied to #[repr(u8)] enums",
+        )
+        .to_compile_error();
+    }
+
+    for (i, variant) in enu.variants.iter().enumerate() {
+        if variant.fields != Fields::Unit {
+            // This can be supported in the future, see zerovec/design_doc.md
+            return Error::new(
+                variant.span(),
+                "#[make_ule] can only be applied to enums with dataless variants",
+            )
+            .to_compile_error();
+        }
+
+        if let Some((_, ref discr)) = variant.discriminant {
+            if let Some(n) = get_expr_int(discr) {
+                // We require explicit discriminants so that it is clear that reordering
+                // fields would be a breaking change. Furthermore, using explicit discriminants helps ensure that
+                // platform-specific C ABI choices do not matter.
+                // We could potentially add in explicit discriminants on the user's behalf in the future, or support
+                // more complicated sets of explicit discriminant values.
+                if n != i as u64 {
```

Comment:
Thought: Forcing the repr of an enum does *not* appear to hinder the compiler's ability to optimize `Option<MyEnum>`.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e7a6abd487bb89f31ccf3d243fd57dfa